### PR TITLE
[gemini] a new tensor structure

### PR DESCRIPTION
**Pull request by feifeibear**
Init a new data structure for tensor.
related to #557

### CODE REVIEW

**Comment by feifeibear** [comment]
image

File: colossalai/gemini/tensor/__init__.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    Example::
+        >>> @custom_stateful_op(torch.nn.functional.linear)
+        >>> def my_custom_linear(types, args, kwargs, process_group):
+        >>>   ....
+        >>>
+        >>> input = torch.rand(10, 32)
+        >>> weight = StatefulTensor(torch.rand(32, 16))
+        >>> bias = StatefulTensor(torch.rand(16))
+        >>> # This will call 'my_custom_stateful_linear'
```
Comment (FrankLeeeee) [inline colossalai/gemini/tensor/__init__.py:22]: I couldn't find `my_custom_stateful_linear`

File: colossalai/gemini/tensor/__init__.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+def stateful_op_impl(func):
+    """
+    Provides a way for users to write their own custom operator. This
+    can be used to override existing StatefulTensorV2 operators or write a new
+    one not supported by StatefulTensorV2. If the operator in question is covered
+    by ``__torch_function__`` dispatch and has a StatefulTensorV2 as any of its
+    parameters, the function provided will be invoked for that operator.
+
+    Example::
+        >>> @custom_stateful_op(torch.nn.functional.linear)
```
Comment (FrankLeeeee) [inline colossalai/gemini/tensor/__init__.py:15]: should this be `stateful_op_impl`?

File: colossalai/gemini/tensor/stateful_tensor.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    def __init__(self, t: torch.Tensor) -> None:
+        self._torch_tensor = t
+
+    def torch_tensor(self) -> torch.Tensor:
+        return self._torch_tensor
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        global _STATEFUL_OPS
+        print(_STATEFUL_OPS)
```
Comment (FrankLeeeee) [inline colossalai/gemini/tensor/stateful_tensor.py:19]: remove this line?

File: colossalai/gemini/tensor/__init__.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/gemini/tensor/__init__.py:15]: Yes

File: colossalai/gemini/tensor/stateful_tensor.py (hunk 3 above)
Comment (feifeibear) [reply colossalai/gemini/tensor/stateful_tensor.py:19]: OK

**Comment by FrankLeeeee** [comment]
What is the purpose of state op?

**Comment by feifeibear** [comment]
hijack the original OP. You do not need to modify Linear as col.nn.Linear

**Comment by FrankLeeeee** [comment]
> hijack the original OP. You do not need to modify Linear as col.nn.Linear

Cool.

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**